docs/git-guide.md

```diff
@@ -0,0 +1,1052 @@
+$ git remote add <username> https://github.com/<username>/zulip.git
+$ git fetch <username>
+```
+
+Now you can checkout their branch just like you would any other. You can name the branch anything you want, but using both the username and branch name will help you keep things organized.
+
+```
+$ git checkout -b <username>-<branchname>
+```
+
+### Checkout a pull request locally
+
+Just as you can checkout any user's branch locally, you can also checkout any pull request locally. GitHub provides a special syntax ([details][github-help-co-pr-locally]) for this since pull requests are specific to GitHub rather than Git.
+
+First, fetch and create a branch for the pull request, replacing *ID* and *BRANCHNAME* with the ID of the pull request and your desired branch name:
+
+```
+$ git fetch upstream pull/ID/head:BRANCHNAME
+```
+
+Now switch to the branch:
+
+```
+$ git checkout BRANCHNAME
+```
+
+Now you work on this branch as you would any other.
+
+## Review changes
+
+### Changes on (local) working tree
+
+Display changes between index and working tree (what is not yet staged for commit):
+
+```
+$ git diff
+```
+
+Display changes between index and last commit (what you have staged for commit):
+
+```
+$ git diff --cached
+```
+
+Display changes in working tree since last commit (changes that are staged as well as ones that are not):
+
+```
+$ git diff HEAD
+```
+
+### Changes within branches
+
+Use any git-ref to compare changes between two commits on the current branch.
+
+Display changes between commit before last and last commit:
+
+```
+$ git diff HEAD^ HEAD
+```
+
+Display changes between two commits using their hashes:
+
+```
+$ git diff e2f404c 7977169
+```
+
+### Changes between branches
+
+Display changes between tip of topic branch and tip of master branch:
+
+```
+$ git diff topic master
+```
+
+Display changes that have occured on master branch since topic branch was created:
+
+```
+$ git diff topic...master
+```
+
+## Get out of trouble
+
+Git is a powerful yet complex version control system. Even for contributors experienced at using version control, it can be confusing. The good news is that nearly all Git actions add information to the Git database, rather than removing it. As such, it's hard to make Git perform actions that you can't undo. However, git can't undo what it doesn't know about, so it's a good practice to frequently commit your changes and freqently push your commits to your remote repository.
+
+### Undo a merge commit
+
+A merge commit happens when the last commit on your current branch is not a direct ancestor of the branch you are trying to merge in. This happens quite often in a busy project like Zulip where there are many contributors because upstream/zulip will have new commits while you're working on a feature or bugfix. In order for Git to merge your changes and the changes that have occured on zulip/upstream since you first started your work, it must perform a three-way merge and create a merge commit.
```

Comment:
This explanation doesn't seem quite right, in that it makes it feel like these happen automatically, when in fact they're created by those commands (this explanation seems a bit mixed up with merge conflicts).  I'd probably mention the definition of a merge commit as this cool thing that lets you create a commit with 2 parents, but that we don't use them.